### PR TITLE
Bicep markdown documentation

### PR DESCRIPTION
**Pull request by Kaloszer**
# Contributing a Pull Request

## Overview/Summary

Based on the work in #224 
Fixes:
- #292 
- #223

## Tests

Some pester tests have been added, they are mostly regex matching, so with time they might break, and adjustment might be needed. But they do their work and validate that the correct items have been created in the markdown file.

Also parts of the code that would create each headers' contents have been split into their own functions with their own test files, if some edge case comes to mind test cases can be added.

There's also another cmdlet added which uses regex to parse module names and their paths.

I have tried to rebase this from main, due to that there's some files that '''changed''' but I believe they're in sync with PSBicep/PSBicep main.

## Note

The actual output of the cmdlet has 2 '#' not 4 as it is in this description, but for PR aesthetics' sake I changed it manually here.

## This PR fixes/adds/changes/removes

1. Generates a Markdown document which contains the following

### Metadata
Metadata can be provided in the bicep file in 2 ways.

``` bicep
metadata pesterMeta = {
  author: 'PSBicep'
  version: 'v1.0'
  aNumber: 1
}

metadata value = 'something'
metadata someothervalue = 'something else'
```

Both are handled and output is as follows:

#### Metadata

| Name | Value |
|----|----|
| pesterMeta | aNumber: 1<br/> author: PSBicep<br/> version: v1.0<br/> |
| someothervalue | something else |
| value | something |
| _generator | name: bicep<br/> templateHash: 13742267763140291220<br/> version: 0.11.1.770<br/> |

### Providers

Type, Version is provided

#### Providers

| Type | Version |
|----|----|
| Microsoft.Authorization/roleAssignments | 2020-04-01-preview |
| Microsoft.Resources/deployments | 2020-10-01 |

### Resources

Name, Link, Location is provided

#### Resources

| Name | Link | Location |
|----|----|----|
| sentinel-module | [Microsoft.Resources/deployments@2020-10-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Resources/2020-10-01/deployments?tabs=bicep) |  |
| apiOutputHelper | [Microsoft.Resources/deployments@2020-10-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Resources/2020-10-01/deployments?tabs=bicep) |  |
| dataConnectors-module | [Microsoft.Resources/deployments@2020-10-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Resources/2020-10-01/deployments?tabs=bicep) |  |

### Parameters

Name, Type, AllowedValues (@allowed decorator), Metadata (@description/other decorators) 
Each item is 'newlined' so they are in a single table cell.

#### Parameters

| Name | Type | AllowedValues | Metadata |
|----|----|----|----|
| customerName | string | n/a | description: Customer name<br/> |
| dataConnectors | array | n/a | description: Data connectors to be enabled in Azure Sentinel<br/> |
| sentinelRetentionInDays | int | 30 <br/> 60 <br/> 90 <br/> 120 <br/> 150 <br/> 180 <br/> 365 <br/> 730 <br/> 1825 <br/> | description: Azure Sentinel Retention in days<br/> |

### Variables

Name, Value is provided (which can be a result of a method so it's a little bit ugly sometimes) ex:

#### Variables
| Name | Value |
|----|----|
| suffix | [parameters('customerName')] |
| startIndexes | [filter(range(0, variables('analyticRuleCount')), lambda('item', equals(mod(lambdaVariables('item'), variables('maxARDeploymentSize')), 0)))] |

###  Outputs

Name, Type, Value is provided

### Modules

Name, Path is provided

#### Modules

| Name | Path |
|----|----|
| managedIdentityModule | ./modules/resources/managedIdentity.bicep |

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/PSBicep/PSBicep/pulls)
- [x] Associated it with relevant [issues](https://github.com/PSBicep/PSBicep/issues), for tracking and closure.
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/PSBicep/PSBicep/tree/main)
- [x] Performed testing.
- [ ] Verified build scripts work.
- [ ] Updated relevant and associated documentation.

### CODE REVIEW

**Comment by StefanIvemo** [comment]
@Kaloszer is this ready for review?

**Comment by Kaloszer** [comment]
@StefanIvemo - I'd say leave it for a couple of days until I finish implementing this in my project, then I'll be certain it works as intended. I'll try to also add a sample ADO .yaml on how to implement it. 

Question is, should I create an root/Examples/.azuredevops folder in the project or root/.azuredevops?

**Comment by Kaloszer** [comment]
@StefanIvemo 

Ready for review.

'works on my machine'™

**Comment by StefanIvemo** [comment]
@Kaloszer we will review this PR shortly. We've been focusing on shipping a pre-release version of the module using the latest Bicep version and getting up to par with new Bicep features. Thank you for your patience.

File: Source/Public/New-BicepMarkdownDocumentation.ps1 [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        [switch]$Console,
+
+        [Parameter(ParameterSetName = 'FromFile')]
+        [Parameter(ParameterSetName = 'FromFolder')]
+        [switch]$Force
+    )
+
+    switch ($PSCmdLet.ParameterSetName) {
+        'FromFile' { 
+            $FileCollection = @((Get-Item $File))
```
Comment (bjompen) [inline Source/Public/New-BicepMarkdownDocumentation.ps1:24]: Instead of forcing Get-Item to return an array, would it maybe make more sense to typecast $FileCollection to an array in both FromFile and FromFolder instead? this may lessen the risk of erroneous types. `[System.IO.FileInfo[]]$FileCollection =`

**Comment by Kaloszer** [comment]
I'll take a look on these over the weekend 👍🏻

File: Source/Public/New-BicepMarkdownDocumentation.ps1 (hunk 1 above)
Comment (Kaloszer) [reply Source/Public/New-BicepMarkdownDocumentation.ps1]: Changed, actually never seemed to have this sort of issue, I think.

File: Source/Public/Get-UsedModulesInBicepFile.ps1 [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        $BicepFile = Get-Content -Path $Path -Raw
+    }
+    catch {
+        throw "Could not read file $Path"
+    }
+
+    $UsedModules = @()
+
+
+    $BicepFile | Select-String -Pattern "module\s+(\w+)\s+'([^']+)'" -AllMatches | ForEach-Object {
```
[the remark this answers is not in the document]
Comment (Kaloszer) [reply Source/Public/Get-UsedModulesInBicepFile.ps1]: Had to change the method as Select-String wasn't parsing it using this regex. But yes, that would be better. Changed it.

**Comment by SimonWahlin** [comment]
@bjompen can you have a look at the changes please?

**Comment by bjompen** [comment]
Looks good to me, but we need to resolve the merge conflicts..

**Review by bjompen — APPROVED** [review]

**Comment by PalmEmanuel** [comment]
Apologies for the long delay for this PR! I made a merge from main and fixed conflicts and the build issues, but it seems you've archived the repository @Kaloszer, so it's read-only and I can't push the changed.

If you still want it merged you can unarchive the repository and write here when you've done it, and I'll push it to make it ready for a merge. 🙂

**Comment by Kaloszer** [comment]
@PalmEmanuel didn't realize moving it to archive would have repercussions 🗡️ 

It's unarchived now :)

**Comment by PalmEmanuel** [comment]
Seems I also didn't consider that I don't actually have access to push to the PR branch since it's in your repository, so the easiest way is probably for you to do the merge. Running this should work:

```
git merge upstream/markdown-pr-293
git push
```

Note the `upstream` remote, which needs to point to PSBicep/PSBicep. It should do that automatically for forks, but still. @Kaloszer

**Comment by PalmEmanuel** [comment]
Except for the last commit [57afc46](https://github.com/PSBicep/PSBicep/pull/293/commits/57afc464e480e427d1eb9f007b7cfca8164657e6) everything looks good. The tests require it to load the Bicep module and copy the support files, not only the BicepNet module. I resolved it in the merge conflict so you don't have to worry about it.

All tests are passing and the code is up to date with `main`.

Well done and thanks for the patience with this one! 🙏 🚀

**Review by PalmEmanuel — APPROVED** [review]
